include/matx/transforms/fft/fft_fftw.h

```diff
@@ -208,6 +220,280 @@ struct FftFFTWparams_t {
     return params;
   }
 
+/**
+ * Crude hash on FFT to get a reasonably good delta for collisions. This
+ * doesn't need to be perfect, but fast enough to not slow down lookups, and
+ * different enough so the common FFT parameters change
+ */
+struct FftFFTWParamsKeyHash {
+  std::size_t operator()(const FftFFTWParams_t &k) const noexcept
+  {
+    return (std::hash<uint64_t>()(k.n[0])) + (std::hash<uint64_t>()(k.n[1])) +
+           (std::hash<uint64_t>()(k.fft_rank)) +
+           (std::hash<uint64_t>()(k.batch)) + (std::hash<uint64_t>()(k.istride)) + 
+           (std::hash<uint64_t>()(static_cast<uint64_t>(k.dir))) +
+           (std::hash<uint64_t>()(static_cast<uint64_t>(k.is_fp32)));
+  }
+};
+
+/**
+ * Test FFT parameters for equality. Unlike the hash, all parameters must match.
+ */
+struct FftFFTWParamsKeyEq {
+  bool operator()(const FftFFTWParams_t &l, const FftFFTWParams_t &t) const noexcept
+  {
+    return l.n[0] == t.n[0] && l.n[1] == t.n[1] && l.batch == t.batch &&
+           l.dir == t.dir && l.fft_rank == t.fft_rank &&
+           l.is_fp32 == t.is_fp32 && l.in_place == t.in_place &&
+           l.inembed[0] == t.inembed[0] && l.inembed[1] == t.inembed[1] &&
+           l.onembed[0] == t.onembed[0] && l.onembed[1] == t.onembed[1] &&
+           l.istride == t.istride && l.ostride == t.ostride &&
+           l.idist == t.idist && l.odist == t.odist &&
+           l.transform_type == t.transform_type &&
+           l.irank == t.irank && l.orank == t.orank ;
+  }
+};
+
+using fft_fftw_cache_t = std::unordered_map<FftFFTWParams_t, std::any, FftFFTWParamsKeyHash, FftFFTWParamsKeyEq>;
+
+
+/**
+ * Class to manage FFTW initialization and cleanup for single and
+ * double precision transforms.
+ * 
+ */
+class FFTWPlanManager {
+public:
+  static void inline InitFFTWF() {
+    if (!init_fp32_) {
+      int ret = fftwf_init_threads();
+      MATX_ASSERT_STR(ret != 0, matxAssertError, "fftwf_init_threads() failed");
+      init_fp32_ = true;
+    }
+  }
+
+  static void inline InitFFTW() {
+    if (!init_fp64_) {
+      int ret = fftw_init_threads();
+      MATX_ASSERT_STR(ret != 0, matxAssertError, "fftw_init_threads() failed");
+      init_fp64_ = true;
+    }
+  }
+
+  static void IncrementPlanCount() { active_plans_++; }
+
+  static void DecrementPlanCount() {
+    active_plans_--;
+    if (active_plans_ == 0) {
+      if (init_fp32_) {
+          fftwf_cleanup_threads();
+          fftwf_cleanup();
+          init_fp32_ = false;
+      }
+      if (init_fp64_) {
+          fftw_cleanup_threads();
+          fftw_cleanup();
+          init_fp64_ = false;
+      }
+    }
+  }
+
+private:
+  static inline int active_plans_ = 0;
```

Comment:
This won't be thread-safe and will lead to a race condition with multiple threads. You can make it a `cuda::std::atomic<int>` to solve that. Same thing for the other variables.

Reply:
Sure, will update